core/remotes/docker/fetcher.go

```diff
@@ -318,19 +347,108 @@ func (r dockerFetcher) open(ctx context.Context, req *request, mediatype string,
 			// Could use buffer pool here but this case should be rare
 			n, err := io.Copy(io.Discard, io.LimitReader(resp.Body, offset))
 			if err != nil {
+				if r.config.Limiter != nil {
+					r.config.Limiter.Release(parallelism)
+				}
 				return nil, fmt.Errorf("failed to discard to offset: %w", err)
 			}
 			if n != offset {
+				if r.config.Limiter != nil {
+					r.config.Limiter.Release(parallelism)
+				}
 				return nil, errors.New("unable to discard to offset")
 			}
 
+			// content range ignored, we can't do concurrent fetches here.
+			if parallelism > 1 && r.config.Limiter != nil {
+				r.config.Limiter.Release(parallelism - 1)
+			}
+			parallelism = 1
 		}
 	}
 
 	body := resp.Body
 	encoding := strings.FieldsFunc(resp.Header.Get("Content-Encoding"), func(r rune) bool {
 		return r == ' ' || r == '\t' || r == ','
 	})
+
+	totalSize, _ := strconv.ParseInt(resp.Header.Get("Content-Length"), 10, 0)
+	remaining := totalSize - offset
+	if parallelism > 1 && chunkSize > 0 && remaining > minimumSizeForParallelDL && req.body == nil {
+		// If we have a content length, we can use multiple requests to fetch
+		// the content in parallel. This will make download of bigger bodies
+		// faster, at the cost of parallelism more requests and max
+		// ~(max_parallelism * goroutine footprint) memory usage. The goroutine
+		// footprint should be: the goroutine stack + copy buffer size
+		numChunks := remaining / chunkSize
+		if numChunks*chunkSize < remaining {
+			numChunks++
+		}
+		queue := make(chan int64, parallelism)
+		stopChan := make(chan struct{})
+		readers, writers := make([]io.Reader, numChunks), make([]*io.PipeWriter, numChunks)
+		for i := int64(0); i < numChunks; i++ {
+			readers[i], writers[i] = io.Pipe()
```

Comment:
Did you happen to evaluate other mechanism to download chunks? Since this uses io.pipe, it was somewhat slower on my tests since writers were all paused until readers read those chunks in order. I was able to improve the performance significantly (will post numbers shortly) by storing all these chunks in a temp directory and reading them after all downloads are complete. 

```
                tempDir, _ := os.MkdirTemp("/tmp", "digest")
		var wg sync.WaitGroup
		for p := int64(0); p < parallelism; p++ {
			wg.Add(1)
			// start parallel download workers
			...
					if i == 0 {
						_, err = io.Copy(f, io.LimitReader(resp.Body, chunkSize))
						_ = resp.Body.Close()
						continue
					}
					...
					_, err = io.Copy(f, nresp.Body)
					_ = nresp.Body.Close()
					...
			}()
		}
		wg.Wait()
		body = getSortedPartsReader(tempDir)
		...
		
func getSortedPartsReader(tempDir string) io.ReadCloser {
	matches, _ := filepath.Glob(filepath.Join(tempDir, "*.part*"))
	c := collate.New(language.English, collate.Numeric)
	c.SortStrings(matches)
	readers := make([]io.Reader, len(matches))
	for i, filename := range matches {
		file, _ := os.Open(filename)
		readers[i] = file
	}
	return io.NopCloser(io.MultiReader(readers...))
}
```

Reply:
I also had a much faster test that was downloading in a file with punched holes, that was used as a buffer.
I like your idea to use parts like this too, but this feels like adding a new set of feature to download into a tmp dir.

Would that be okay to do that?
Should we cleanup the folder afterwards ?

Reply:
We could add a part to auto remove files onces downloaded. We will have to be careful with files that don't exist yet as `getSortedPartsReader` will list what is currently there, but not the rest.

Reply:
I personally feel like we should merge this PR, and then I will try to do the punched-buffer-file thing.